### PR TITLE
perf(reconciler): skip redundant automation-name write when value already matches (P3)

### PR DESCRIPTION
**Pull request by azchohfi**
## P3 — skip the redundant per-cell automation-name write

**Phase-1 finding.** `Reconciler.UpdateDefaultAutomationName` runs on every changed cell that goes through `Update`. In the steady state the caption-derived default it computes already equals the live UIA Name, so the `SetName` it issues is a value no-op — pure per-cell overhead.

### Change
- Extract the decision into a **pure, DP-free** `ResolveDefaultAutomationNameUpdate(current, oldCaption, newCaption)` helper (headless-testable in `Reactor.Tests`).
- Add an **idempotent-write guard**: after computing the trimmed caption-derived default, skip the `SetName` when the live Name already equals it. `GetName` still runs for non-whitespace captions — **only the redundant same-value write is skipped**.
- (Behavior-identical micro-skip: a whitespace new caption returns before the `GetName` read, exactly as main's first line does.)

### Correctness (MED-risk area — automation names)
The helper models main's `GetName` + author-override + `SetName` logic **exactly**, plus the idempotent guard:
- Author override (live Name ≠ old caption) → left untouched. ✅
- Genuine caption change A→B → writes B. ✅
- **Restore default**: a removed `.AutomationName()` override makes `ApplyModifiers` clear the Name *before* this runs; with an unchanged caption the default is still re-applied (empty `current` ≠ trimmed → write). ✅ *(This is the regression a naïve "skip when caption unchanged" cut would have introduced — caught in review, now pinned by a teeth test.)*
- Steady state (live Name already == default) → skip the redundant write. ✅ *(the saving)*
- `>100`-char trim preserved. ✅

### Tests
`ReconcilerAutomationNameTests` (12 cases) pin the decision with **teeth both ways**:
- Revert the idempotent guard → the skip test sees a redundant `"X"` write → fails.
- Re-add a blanket unchanged-caption skip → the restore-default test returns `null` → fails.

### Scope / gate
- **`src/Reactor/Core/Reconciler.cs` only** — file-disjoint from #692 (P1) and the rest of the perf fleet.
- AOT core-lib build **0W/0E**; full `Reactor.Tests` **9714 passed / 0 failed / 64 skipped**.
- **Honest framing:** #692 already removed the big alloc chunk. On StocksGrid specifically the guard rarely fires (changed cells have *changed* captions), so the win is a modest base-cost shave and may read within-noise under `/perf` — that's fine *data* (the automation write isn't a StocksGrid bottleneck). The guard still removes provably-redundant writes for any control whose non-caption props change under an unchanged caption.

🔒 **DRAFT + merge-gated:** does not merge until the improved `/perf` can resolve its diff **and** the user gives explicit GO (same rule as #692).

### CODE REVIEW

File: src/Reactor/Core/Reconciler.cs [hunk 1]
```diff
@@ -2683,14 +2683,37 @@ public static void ApplyDefaultAutomationName(FrameworkElement fe, string? capti
...
+        // nothing to do). These two arms mirror the caption-only arms of
+        // ResolveDefaultAutomationNameUpdate, so the live path and the pure
+        // decision helper stay in lockstep. On a changed caption we fall through to
+        // the original GetName + author-override + SetName behavior, unchanged.
         if (string.IsNullOrWhiteSpace(newCaption)) return;
+        if (string.Equals(oldCaption, newCaption, StringComparison.Ordinal)) return;
         var current = Microsoft.UI.Xaml.Automation.AutomationProperties.GetName(fe);
+        var resolved = ResolveDefaultAutomationNameUpdate(current, oldCaption, newCaption);
+        if (resolved is not null)
+            Microsoft.UI.Xaml.Automation.AutomationProperties.SetName(fe, resolved);
```
[the remark this answers is not in the document]
Comment (azchohfi) [reply src/Reactor/Core/Reconciler.cs:2699]: Good catch — fixed in 6416cb45. You're right that `ApplyModifiers` runs before this and can clear `NameProperty` when an explicit `.AutomationName()` override is removed, so a blanket `oldCaption == newCaption` skip drops the restore-default case.

Replaced it with an idempotent-write guard: I keep main's `GetName` + author-override logic verbatim, compute the trimmed target, and skip **only** the `SetName` when the live Name already equals it (a value no-op). The cleared-Name + unchanged-caption case now falls through and writes the caption default — identical to main. Added a teeth test `Cleared_Name_Restores_Caption_Default_When_Caption_Unchanged` pinning exactly this.

File: src/Reactor/Core/Reconciler.cs [hunk 2]
```diff
@@ -2683,14 +2683,37 @@ public static void ApplyDefaultAutomationName(FrameworkElement fe, string? capti
...
+    // Pure decision for UpdateDefaultAutomationName — no DP/UIA interop, so the
+    // caption/override policy is unit-testable headlessly (Reactor.Tests). Returns
+    // the Name to write, or null to leave the live automation Name untouched. Kept
+    // in sync with the UpdateDefaultAutomationName fast-path above: the whitespace-
+    // new-caption and unchanged-caption arms here are the same skips the live path
+    // takes before it ever reads the current Name.
+    internal static string? ResolveDefaultAutomationNameUpdate(string? current, string? oldCaption, string? newCaption)
+    {
+        if (string.IsNullOrWhiteSpace(newCaption)) return null;
+        if (string.Equals(oldCaption, newCaption, StringComparison.Ordinal)) return null;
```
[the remark this answers is not in the document]
Comment (azchohfi) [reply src/Reactor/Core/Reconciler.cs]: Done in 6416cb45 — the helper no longer special-cases unchanged captions. It now mirrors main's author-override logic and only short-circuits via the idempotent-write guard (`current == trimmed`), so an empty `current` with an unchanged caption resolves to the caption default (a write), not `null`.

File: tests/Reactor.Tests/ReconcilerAutomationNameTests.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    [Fact]
+    public void Unchanged_Caption_Returns_Null_Even_When_Live_Name_Is_Empty()
+    {
+        // TEETH for P3: with the `oldCaption == newCaption` fast-path the unchanged caption is a
+        // skip regardless of the live Name. Remove that line and the helper falls through to the
+        // author-override logic: current "" is not an override, so it returns "X" (a write) and
+        // this assertion fails — i.e. reverting the optimization breaks this test.
+        Assert.Null(Reconciler.ResolveDefaultAutomationNameUpdate(current: "", oldCaption: "X", newCaption: "X"));
+    }
```
[the remark this answers is not in the document]
Comment (azchohfi) [reply tests/Reactor.Tests/ReconcilerAutomationNameTests.cs:44]: Agreed — rewritten in 6416cb45. That case now asserts a write (`Assert.Equal("X", ...)`) to pin the restore-default behavior, and a separate teeth test pins the idempotent-write skip. Reverting either direction fails a test.

File: src/Reactor/Core/Reconciler.cs [hunk 4]
```diff
@@ -2683,14 +2683,40 @@ public static void ApplyDefaultAutomationName(FrameworkElement fe, string? capti
     public static void UpdateDefaultAutomationName(FrameworkElement fe, string? oldCaption, string? newCaption)
     {
         if (fe is null) return;
+        // A whitespace new caption has nothing to write — skip the GetName read too
+        // (main's first line returns on whitespace newCaption, so this is behavior-
+        // identical, it just avoids the interop read). Otherwise read the live Name
+        // and let the pure helper decide; we only touch the DP when it returns a value.
```
[the remark this answers is not in the document]
Comment (azchohfi) [reply src/Reactor/Core/Reconciler.cs]: Clarified in 31302db0 — reworded the comment to make explicit that `GetName` still runs for non-whitespace captions and only the redundant `SetName` is skipped. Also updated the PR title/description to describe the idempotent-write guard rather than a "round-trip skip". Thanks for the precision.

File: src/Reactor/Core/Reconciler.cs [hunk 5]
```diff
@@ -2683,14 +2683,44 @@ public static void ApplyDefaultAutomationName(FrameworkElement fe, string? capti
...
     {
         if (fe is null) return;
+        // Two arms: a whitespace new caption has nothing to write, so we return before
+        // even reading the Name (main also returns first on whitespace newCaption, so
+        // this is behavior-identical — it just avoids that one interop read). For any
+        // NON-whitespace caption we still read the live Name and let the pure helper
+        // decide. The actual P3 saving lives inside that helper: it skips the SetName
+        // write when the live Name already equals the caption-derived default (the
+        // redundant same-value write). GetName is NOT skipped for unchanged captions —
+        // only the redundant SetName is. We touch the DP only when the helper returns a value.
```
[the remark this answers is not in the document]
Comment (azchohfi) [reply src/Reactor/Core/Reconciler.cs]: Reworded in 3ef87ebd. The comment now scopes the conditional DP access to the SetName *write* and notes the GetName read always runs for a non-whitespace caption — so it no longer implies the DP is otherwise untouched. Comment-only; no behavior change.

**Comment by azchohfi** [comment]
Closing after re-validation on current main (post-#692 + #665). Confirmed an honest no-op: automation-name-churn is below resolution (positional memoized cells rarely re-set an identical allocating automation name); all macro blocks flat, all 17 micros 0.0% deterministic. Reopen if an automation-name-churn micro/workload is added.
